### PR TITLE
[Feature] AVS error alert

### PR DESCRIPTION
**Pull request by johnxnguyen**
## What's in this PR?

In preparation for conference calls, we need to handle the case where an older client receives a conference call. Since the old client can not connect to the call, we display an alert dialog informing the user to update the client.

When we receive an incoming call message, we pass it along to AVS. If AVS detects that it is a conference call and the current client version does not support it, it will return an error code `1000` corresponding to `WCALL_ERROR_UNKNOWN_PROTOCOL`. When this happens, we need to propagate the event up to the UI and display the alert.

There are two cases in which the alert is shown:

1. If the app is in the foreground, it will be displayed as soon as the error occurs.
2. If the app is in the background, it will be displayed as soon as the app comes back to the foreground.
#### APK
[Download build #2171](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2171/artifact/build/artifact/wire-dev-PR2862-2171.apk)
[Download build #2172](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2172/artifact/build/artifact/wire-dev-PR2862-2172.apk)

### CODE REVIEW

File: wire-android-sync-engine/zmessaging/src/main/scala/com/waz/service/call/CallingServiceImpl.scala [hunk 1]
```diff
@@ -527,7 +529,9 @@ class CallingServiceImpl(val accountId:       UserId,
       val drift = pushService.beDrift.currentValue.getOrElse(Duration.ZERO)
       val curTime = LocalInstant(clock.instant + drift)
       verbose(l"Received msg for avs: localTime: ${clock.instant} curTime: $curTime, drift: $drift, msgTime: $msgTime")
-      avs.onReceiveMessage(w, msg, curTime, msgTime, convId, from, sender)
+      avs.onReceiveMessage(w, msg, curTime, msgTime, convId, from, sender).foreach { result =>
+        callError ! result
```
Comment (johnxnguyen) [inline wire-android-sync-engine/zmessaging/src/main/scala/com/waz/service/call/CallingServiceImpl.scala:533]: @makingthematrix I know this is getting called, but my subscribers don't get notified...

File: app/src/main/scala/com/waz/zclient/calling/controllers/CallController.scala [hunk 2]
```diff
@@ -188,6 +188,10 @@ class CallController(implicit inj: Injector, cxt: WireContext, eventContext: Eve
       c  <- callConvId
     } yield (cs, c)
 
+  val onCallError = callingService
+    .flatMap(_.callError)
+    .onChanged
+    .filter { _ != AvsCallError.None }
```
Comment (johnxnguyen) [inline app/src/main/scala/com/waz/zclient/calling/controllers/CallController.scala:194]: This is what we will observe from the UI. We're just interested when an error occurs.

File: app/src/main/scala/com/waz/zclient/MainActivity.scala [hunk 3]
```diff
@@ -225,6 +227,19 @@ class MainActivity extends BaseActivity
...
+        new AlertDialog.Builder(this)
+          .setTitle(R.string.call_error_unsupported_version_title)
+          .setMessage(R.string.call_error_unsupported_version_message)
+          .setPositiveButton(R.string.call_error_unsupported_version_button_ok, new DialogInterface.OnClickListener {
+            override def onClick(dialog: DialogInterface, which: AvsCallError): Unit = inject[BrowserController].openPlayStoreListing()
+          })
+          .setNegativeButton(R.string.call_error_unsupported_version_button_dismiss, null)
+          .show()
+      case _ =>
+    }
```
Comment (johnxnguyen) [inline app/src/main/scala/com/waz/zclient/MainActivity.scala:242]: Neither of these cases get called... I'm not sure why.

File: wire-android-sync-engine/zmessaging/src/main/scala/com/waz/service/call/CallingServiceImpl.scala [hunk 4]
```diff
@@ -527,7 +527,9 @@ class CallingServiceImpl(val accountId:       UserId,
       val drift = pushService.beDrift.currentValue.getOrElse(Duration.ZERO)
       val curTime = LocalInstant(clock.instant + drift)
       verbose(l"Received msg for avs: localTime: ${clock.instant} curTime: $curTime, drift: $drift, msgTime: $msgTime")
-      avs.onReceiveMessage(w, msg, curTime, msgTime, convId, from, sender)
+      avs.onReceiveMessage(w, msg, curTime, msgTime, convId, from, sender).foreach { result =>
+        userPrefs(UserPreferences.ShouldWarnAVSUpgrade).mutate(_ | result == AvsCallError.UnknownProtocol)
```
Comment (makingthematrix) [inline wire-android-sync-engine/zmessaging/src/main/scala/com/waz/service/call/CallingServiceImpl.scala:531]: clever :)

File: app/src/main/scala/com/waz/zclient/MainActivity.scala [hunk 5]
```diff
@@ -225,6 +225,20 @@ class MainActivity extends BaseActivity
...
+          showAVSUpgradeWarning(accentColor) { didConfirm =>
+            if (didConfirm) inject[BrowserController].openPlayStoreListing()
+            userPreferences.head.foreach { prefs =>
+              prefs(UserPreferences.ShouldWarnAVSUpgrade) := false
+            }
+          }
+        }
+      }
+    }
+
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/MainActivity.scala:241]: Looks like readability could be improved, but let it be for now.

**Pull request merged**